File: dominoes.rs
use std::collections::HashSet;

pub fn chain(input: &[(u8, u8)]) -> Option<Vec<(u8, u8)>> {
    if input.len() == 0 {return Some(Vec::new());}
    let mut rever = Vec::new();
    for (start, end) in input {
        rever.push((*end, *start));
    }
    for i in 0..input.len() {
        let mut vis: HashSet<usize> = HashSet::new();
        vis.insert(i);
        let mut vec = Vec::new();
        vec.push((input[i].0,input[i].1));
        let chain = f(input[i].0, input[i].1, input,&rever, &mut vis,&mut vec);
        if chain.is_some(){return chain;}
    }
    None
}
fn get_candidate(candidate_idx : &mut Vec<usize>,input: &[(u8, u8)],end: u8,vis: &HashSet<usize>){
    let mut idx = 0;
    for (start,_) in input{
        if *start == end && !vis.contains(&idx){candidate_idx.push(idx);}
        idx += 1;
    }
}
fn f(first: u8, end: u8, input: &[(u8, u8)],rever : &Vec<(u8,u8)>, vis: &mut HashSet<usize>,vec: &mut Vec<(u8,u8)>) -> Option<Vec<(u8, u8)>> {
    if vis.len() == input.len(){
        if first == end && is_right(vec){return Some(vec.to_vec());}
        return None;
    }
    let mut candidate_idx_input = Vec::new();
    let mut candidate_idx_rever = Vec::new();
    get_candidate(&mut candidate_idx_input,input,end,vis);
    get_candidate(&mut candidate_idx_rever,rever,end,vis);
    let mut first_ans ;
    for idx in candidate_idx_input{
        vec.push(input[idx]);
        vis.insert(idx);
        first_ans = f(first, input[idx].1, input, rever, vis, vec);
        if first_ans.is_some(){return first_ans;}
        vis.remove(&idx);
        vec.pop();
    }
    first_ans = None;
    for idx in candidate_idx_rever{
        vec.push(rever[idx]);
        vis.insert(idx);
        first_ans = f(first, rever[idx].1, input, rever, vis, vec);
        if first_ans.is_some(){return first_ans;}
        vis.remove(&idx);
        vec.pop();
    }
    first_ans
}
fn is_right(input: &Vec<(u8, u8)>) -> bool {
    let (first, mut equal) = input[0];
    for i in 1..input.len() {
        let (f, e) = input[i];
        if equal != f {return false;}
        equal = e;
    }
    if equal != first {return false;}
    true
}
// PaulDance's solution is so much better.. check that out.
// concept - Eulerian cycle in graph

File: solutions/python/dominoes/1/dominoes.py
from typing import List, Tuple, Optional, Set
def can_chain(input: List[Tuple[int, int]]) -> Optional[List[Tuple[int, int]]]:
    if len(input) == 0:
        return []
    rever = [(end, start) for start, end in input]
    for i in range(len(input)):
        vis: Set[int] = set()
        vis.add(i)
        vec = [input[i]]
        result = f(input[i][0], input[i][1], input, rever, vis, vec)
        if result is not None:
            return result
    return None


def get_candidate(candidate_idx: List[int], input: List[Tuple[int, int]], end: int, vis: Set[int]):
    for idx, (start, _) in enumerate(input):
        if start == end and idx not in vis:
            candidate_idx.append(idx)


def f(first: int, end: int,
      input: List[Tuple[int, int]],
      rever: List[Tuple[int, int]],
      vis: Set[int],
      vec: List[Tuple[int, int]]) -> Optional[List[Tuple[int, int]]]:
    if len(vis) == len(input):
        if first == end and is_right(vec):
            return vec.copy()
        return None
    candidate_idx_input = []
    candidate_idx_rever = []
    get_candidate(candidate_idx_input, input, end, vis)
    get_candidate(candidate_idx_rever, rever, end, vis)

    for idx in candidate_idx_input:
        vec.append(input[idx])
        vis.add(idx)
        result = f(first, input[idx][1], input, rever, vis, vec)
        if result is not None:
            return result
        vis.remove(idx)
        vec.pop()

    for idx in candidate_idx_rever:
        vec.append(rever[idx])
        vis.add(idx)
        result = f(first, rever[idx][1], input, rever, vis, vec)
        if result is not None:
            return result
        vis.remove(idx)
        vec.pop()
    return None


def is_right(input: List[Tuple[int, int]]) -> bool:
    first, equal = input[0]
    for i in range(1, len(input)):
        f, e = input[i]
        if equal != f:
            return False
        equal = e
    return equal == first